### PR TITLE
[history-tracker] track changes to router table (next hop, cost)

### PR DESCRIPTION
**Pull request by abtink**
This commit adds anew feature in `HistoryTracker` to track the history of changes to routers. In particular we track when a new router is added or removed, or when next hop and/or path cost towards a router is changed. On a cost change the recorded entry, provides the old cost value along with the new path cost. This helps to track the cost changes more easily (no need to look back at history to find the old cost value).

This commit also adds CLI command `history router` for the new feature.

### CODE REVIEW

File: include/openthread/history_tracker.h [hunk 1]
```diff
@@ -192,6 +192,33 @@ typedef struct otHistoryTrackerNeighborInfo
...
 } otHistoryTrackerNeighborInfo;
 
+/**
+ * This enumeration defines the events in a router info (i.e. whether router is added, removed, or changed).
+ *
+ */
+typedef enum
+{
+    OT_HISTORY_TRACKER_ROUTER_EVENT_ADDED            = 0, ///< Router is added (router ID allocated).
+    OT_HISTORY_TRACKER_ROUTER_EVENT_REMOVED          = 1, ///< Router  entry is removed (router ID is released).
```
Comment (jwhui) [inline include/openthread/history_tracker.h:202]:
```suggestion
    OT_HISTORY_TRACKER_ROUTER_EVENT_REMOVED          = 1, ///< Router entry is removed (router ID is released).
```

File: src/cli/README_HISTORY.md [hunk 2]
```diff
@@ -337,6 +339,66 @@ Print the history as a list (last two entries).
 Done
 ```
 
+### router
+
+Usage `history router [list] [<num-entries>]`
+
+Print the route table history. Each item provides:
+
+- Event (`Added`, `Removed`, 'NextHopChnaged', 'CostChanged')
```
Comment (jwhui) [inline src/cli/README_HISTORY.md:348]:
```suggestion
- Event (`Added`, `Removed`, `NextHopChnaged`, `CostChanged`)
```

File: src/cli/README_HISTORY.md [hunk 3]
```diff
@@ -337,6 +339,66 @@ Print the history as a list (last two entries).
...
+- Next Hop (Router ID and RLOC16)
+- Path cost (old `-->` new)
+
+Print the history as a table.
+
+```bash
+> history router
+| Age                  | Event          | ID (RLOC16) | Next Hop    | Path Cost |
++----------------------+----------------+-------------+-------------+-----------+
+|         00:02:53.501 | NextHopChanged | 19 (0x4c00) |           - |         - |
```
Comment (jwhui) [inline src/cli/README_HISTORY.md:359]: Indicates next hop changed, but no change in path cost?

Comment (abtink) [reply src/cli/README_HISTORY.md:359]: `-` means we have no next hop and therefore no path cost as well (showing it as `-`).

Comment (jwhui) [reply src/cli/README_HISTORY.md:359]: Is it worth showing what the old path cost was?

Comment (abtink) [reply src/cli/README_HISTORY.md:359]: Good point. Will update it to show the change.

Comment (abtink) [reply src/cli/README_HISTORY.md]: Updated this in new push.
Another thing I changed is in new push is to output `inf` (instead of `0`) for path cost.

**Review by jwhui — APPROVED** [review]

**Pull request merged**